rat/compiler.py

```diff
@@ -7,82 +7,333 @@
+                record.subscripts |= subscripts
 
-        # this set keep track of variable names that are not parameters, i.e. assigned by assignment
-        self.assigned_parameter_keys: Set[str] = set()
+            if subscript_alias:
+                record.subscript_alias = subscript_alias
 
-        self.data_variables: Dict[str, variables.Data] = {}
-        self.parameter_variables: Dict[str, variables.Param] = {}
-        self.assigned_parameter_variables: Dict[str, variables.AssignedParam] = {}
-        self.subscript_use_variables: Dict[str, variables.SubscriptUse] = {}
-        self.variable_subscripts: Dict[str, variables.Subscript] = {}
-        self.first_in_group_indicators: Dict[str, numpy.ndarray] = {}
+            record.constraint_lower = max(record.constraint_lower, constraint_lower)
+            record.constraint_upper = min(record.constraint_upper, constraint_upper)
 
-    def _get_unique_identifier(self) -> str:
+        else:
+            # Insert if new
+            self.symbol_dict[variable_name] = TableRecord(
+                variable_name,
+                variable_type,
+                subscripts=subscripts if subscripts else set(),
+                subscript_length=len(tuple(subscripts)[0]) if subscripts else 0,
+                subscript_alias=subscript_alias,
+                constraint_lower=constraint_lower,
+                constraint_upper=constraint_upper,
+            )
+
+    def lookup(self, variable_name: str) -> TableRecord:
         """
-        Return an as-yet un-returned string. This is useful for generating unique identifiers
-        for variables
+        Dictionary indexing
         """
-        self._unique_identifier += 1
-        return repr(self._unique_identifier)
+        return self.symbol_dict[variable_name]
+
+    def iter_records(self):
+        for name, record in self.symbol_dict.items():
+            yield name, record
+
+    def build_base_dataframes(self, data_df: pd.DataFrame):
+        """
+        Builds the base dataframes for parameters from the current symbol table.
+        Also resets any generated subscript indices
+        """
+        self.generated_subscript_dict = {}
+        self.generated_subscript_count = 0
+
+        current_index = 0
+        for variable_name, record in self.symbol_dict.items():
+            if not record.variable_type:
+                error_msg = f"Fatal internal error - variable type information for '{variable_name}' not present within the symbol table. Aborting compilation."
+                raise CompileError(error_msg)
+
+            # Data variables don't need a base df; the input df is its base dataframe.
+            if record.variable_type == VariableType.DATA:
+                continue
+
+            if record.subscript_length > 0:
+                base_df = pd.DataFrame()
+
+                # Since the subscripts in the symbol table are all resolved to denote columns in the input dataframe,
```

Comment:
Oh interesting. I though it still would be possible to access a subset of the base dataframe with columns from the input dataframe.
So to clarify, after a variable has been renamed(alias in this PR), it can only be accessed by its new names?

Reply:
Well, the names aren't really used like that so much. Variables are subscripted by subscript names that belong to the primary dataframe.

Like:
```
y' ~ normal(mu[group], sigma);
```

`group` as a name belongs to the dataframe associated with y. `mu` builds itself a dataframe and gives it the name group, but the group name here is still associated with y.